### ISSUE TITLE
FAILED while building

### ISSUE THREAD

**Issue by marchLex** [AOMediaCodec/libavif#1925]
Hi! Hope you'll help me dealing with a problem.

Using ninja I got next error

```
[1/30] Building C object CMakeFiles/avif_obj.dir/src/colrconvert.c.o
FAILED: CMakeFiles/avif_obj.dir/src/colrconvert.c.o
/usr/bin/cc -DAVIF_CODEC_AOM=1 -DAVIF_CODEC_AOM_DECODE=1 -DAVIF_CODEC_AOM_ENCODE=1 -I../include -I../third_party/libyuv/include -isystem ../ext/aom -O2 -DNDEBUG -fPIC -fvisibility=hidden   -Wall -Wextra -Werror -pthread -std=gnu99 -MD -MT CMakeFiles/avif_obj.dir/src/colrconvert.c.o -MF CMakeFiles/avif_obj.dir/src/colrconvert.c.o.d -o CMakeFiles/avif_obj.dir/src/colrconvert.c.o   -c ../src/colrconvert.c
../src/colrconvert.c: In function ‘avifColorPrimariesComputeRGBToXYZD50Matrix’:
../src/colrconvert.c:112:5: error: passing argument 1 of ‘avifVecMul’ from incompatible pointer type [-Werror]
     avifVecMul(rgbPrimariesInv, whitePointXYZ, rgbCoefficients);
     ^
../src/colrconvert.c:76:13: note: expected ‘const double (*)[3]’ but argument is of type ‘double (*)[3]’
 static void avifVecMul(const double M[3][3], const double x[3], double y[3])
             ^
../src/colrconvert.c:118:5: error: passing argument 2 of ‘avifMatMul’ from incompatible pointer type [-Werror]
     avifMatMul(rgbPrimaries, rgbCoefficientsMat, rgbXYZ);
     ^
../src/colrconvert.c:48:13: note: expected ‘const double (*)[3]’ but argument is of type ‘double (*)[3]’
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
../src/colrconvert.c:136:5: error: passing argument 1 of ‘avifMatMul’ from incompatible pointer type [-Werror]
     avifMatMul(adaptation, avifBradford, tmp);
     ^
../src/colrconvert.c:48:13: note: expected ‘const double (*)[3]’ but argument is of type ‘double (*)[3]’
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
../src/colrconvert.c:142:5: error: passing argument 1 of ‘avifMatMul’ from incompatible pointer type [-Werror]
     avifMatMul(bradfordInv, tmp, adaptation);
     ^
../src/colrconvert.c:48:13: note: expected ‘const double (*)[3]’ but argument is of type ‘double (*)[3]’
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
../src/colrconvert.c:142:5: error: passing argument 2 of ‘avifMatMul’ from incompatible pointer type [-Werror]
     avifMatMul(bradfordInv, tmp, adaptation);
     ^
../src/colrconvert.c:48:13: note: expected ‘const double (*)[3]’ but argument is of type ‘double (*)[3]’
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
../src/colrconvert.c:144:5: error: passing argument 1 of ‘avifMatMul’ from incompatible pointer type [-Werror]
     avifMatMul(adaptation, rgbXYZ, coeffs);
     ^
../src/colrconvert.c:48:13: note: expected ‘const double (*)[3]’ but argument is of type ‘double (*)[3]’
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
../src/colrconvert.c:144:5: error: passing argument 2 of ‘avifMatMul’ from incompatible pointer type [-Werror]
     avifMatMul(adaptation, rgbXYZ, coeffs);
     ^
../src/colrconvert.c:48:13: note: expected ‘const double (*)[3]’ but argument is of type ‘double (*)[3]’
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
../src/colrconvert.c: In function ‘avifColorPrimariesComputeXYZD50ToRGBMatrix’:
../src/colrconvert.c:153:5: error: passing argument 1 of ‘avifMatInv’ from incompatible pointer type [-Werror]
     AVIF_CHECK(avifMatInv(rgbToXyz, coeffs));
     ^
../src/colrconvert.c:25:17: note: expected ‘const double (*)[3]’ but argument is of type ‘double (*)[3]’
 static avifBool avifMatInv(const double M[3][3], double I[3][3])
                 ^
../src/colrconvert.c: In function ‘avifColorPrimariesComputeRGBToRGBMatrix’:
../src/colrconvert.c:169:5: error: passing argument 1 of ‘avifMatMul’ from incompatible pointer type [-Werror]
     avifMatMul(xyzToDstRGB, srcRGBToXYZ, coeffs);
     ^
../src/colrconvert.c:48:13: note: expected ‘const double (*)[3]’ but argument is of type ‘double (*)[3]’
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
../src/colrconvert.c:169:5: error: passing argument 2 of ‘avifMatMul’ from incompatible pointer type [-Werror]
     avifMatMul(xyzToDstRGB, srcRGBToXYZ, coeffs);
     ^
../src/colrconvert.c:48:13: note: expected ‘const double (*)[3]’ but argument is of type ‘double (*)[3]’
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
cc1: all warnings being treated as errors
[6/30] Building C object CMakeFiles/avif_obj.dir/src/reformat.c.o
ninja: build stopped: subcommand failed.

```

I'm using CentOS 7. All the libraries are updated. So what a problem could it be?

**Comment by marchLex** [comment]
Is it ok, if I just remove flag [-Werror]?

**Comment by vrabaud** [comment]
This is due to incompatible-pointer-types-discards-qualifiers.
What is weird is that I cannot reproduce it locally with calang, even with `-Wincompatible-pointer-types-discards-qualifiers`.

**Comment by wantehchang** [comment]
Alexander: What is the output of `/usr/bin/cc --version`?

You can pass `-DAVIF_ENABLE_WERROR=OFF` to the cmake command to not treat compiler warnings as errors. The default of the `AVIF_ENABLE_WERROR` option is `ON`.

Vincent: The compiler may be right, but this is beyond my knowledge of the C language.  I only know that in function prototypes, `const double A[3][3]` is the same as `const double A[][3]`. So that matches the `const double (*)[3]` in the warning message.

Rather than trying to understand this esoteric part of the C language, we can put the two-dimensional array in a struct and pass a pointer to the struct as the function parameter.

**Comment by vrabaud** [comment]
A struct would end up having the same problem as it would contain a [3][3]. The simplest I could think of to minimize changes is https://github.com/AOMediaCodec/libavif/pull/1932

**Comment by sezero** [comment]
Hit the same error using gcc-4.4.7 and gcc-4.9.4 in the main branch.

Here is the error output from gcc-4.9.4:

```
/tmp/libavif/src/colrconvert.c: In function 'avifColorPrimariesComputeRGBToXYZD50Matrix':
/tmp/libavif/src/colrconvert.c:112:16: error: passing argument 1 of 'avifVecMul' from incompatible pointer type [-Werror]
     avifVecMul(rgbPrimariesInv, whitePointXYZ, rgbCoefficients);
                ^
/tmp/libavif/src/colrconvert.c:76:13: note: expected 'const double (*)[3]' but argument is of type 'double (*)[3]'
 static void avifVecMul(const double M[3][3], const double x[3], double y[3])
             ^
/tmp/libavif/src/colrconvert.c:118:30: error: passing argument 2 of 'avifMatMul' from incompatible pointer type [-Werror]
     avifMatMul(rgbPrimaries, rgbCoefficientsMat, rgbXYZ);
                              ^
/tmp/libavif/src/colrconvert.c:48:13: note: expected 'const double (*)[3]' but argument is of type 'double (*)[3]'
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
/tmp/libavif/src/colrconvert.c:136:16: error: passing argument 1 of 'avifMatMul' from incompatible pointer type [-Werror]
     avifMatMul(adaptation, avifBradford, tmp);
                ^
/tmp/libavif/src/colrconvert.c:48:13: note: expected 'const double (*)[3]' but argument is of type 'double (*)[3]'
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
/tmp/libavif/src/colrconvert.c:142:16: error: passing argument 1 of 'avifMatMul' from incompatible pointer type [-Werror]
     avifMatMul(bradfordInv, tmp, adaptation);
                ^
/tmp/libavif/src/colrconvert.c:48:13: note: expected 'const double (*)[3]' but argument is of type 'double (*)[3]'
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
/tmp/libavif/src/colrconvert.c:142:29: error: passing argument 2 of 'avifMatMul' from incompatible pointer type [-Werror]
     avifMatMul(bradfordInv, tmp, adaptation);
                             ^
/tmp/libavif/src/colrconvert.c:48:13: note: expected 'const double (*)[3]' but argument is of type 'double (*)[3]'
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
/tmp/libavif/src/colrconvert.c:144:16: error: passing argument 1 of 'avifMatMul' from incompatible pointer type [-Werror]
     avifMatMul(adaptation, rgbXYZ, coeffs);
                ^
/tmp/libavif/src/colrconvert.c:48:13: note: expected 'const double (*)[3]' but argument is of type 'double (*)[3]'
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
/tmp/libavif/src/colrconvert.c:144:28: error: passing argument 2 of 'avifMatMul' from incompatible pointer type [-Werror]
     avifMatMul(adaptation, rgbXYZ, coeffs);
                            ^
/tmp/libavif/src/colrconvert.c:48:13: note: expected 'const double (*)[3]' but argument is of type 'double (*)[3]'
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
In file included from /tmp/libavif/src/colrconvert.c:4:0:
/tmp/libavif/src/colrconvert.c: In function 'avifColorPrimariesComputeXYZD50ToRGBMatrix':
/tmp/libavif/src/colrconvert.c:153:27: error: passing argument 1 of 'avifMatInv' from incompatible pointer type [-Werror]
     AVIF_CHECK(avifMatInv(rgbToXyz, coeffs));
                           ^
/tmp/libavif/include/avif/internal.h:37:15: note: in definition of macro 'AVIF_CHECK'
         if (!(A)) {             \
               ^
/tmp/libavif/src/colrconvert.c:25:17: note: expected 'const double (*)[3]' but argument is of type 'double (*)[3]'
 static avifBool avifMatInv(const double M[3][3], double I[3][3])
                 ^
/tmp/libavif/src/colrconvert.c: In function 'avifColorPrimariesComputeRGBToRGBMatrix':
/tmp/libavif/src/colrconvert.c:169:16: error: passing argument 1 of 'avifMatMul' from incompatible pointer type [-Werror]
     avifMatMul(xyzToDstRGB, srcRGBToXYZ, coeffs);
                ^
/tmp/libavif/src/colrconvert.c:48:13: note: expected 'const double (*)[3]' but argument is of type 'double (*)[3]'
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
/tmp/libavif/src/colrconvert.c:169:29: error: passing argument 2 of 'avifMatMul' from incompatible pointer type [-Werror]
     avifMatMul(xyzToDstRGB, srcRGBToXYZ, coeffs);
                             ^
/tmp/libavif/src/colrconvert.c:48:13: note: expected 'const double (*)[3]' but argument is of type 'double (*)[3]'
 static void avifMatMul(const double A[3][3], const double B[3][3], double C[3][3])
             ^
cc1: all warnings being treated as errors
```

**Comment by vrabaud** [comment]
@sezero , does https://github.com/AOMediaCodec/libavif/pull/1946 fully fixes it for you? Thx

**Comment by sezero** [comment]
> @sezero , does #1946 fully fixes it for you? Thx

Unfortunately no.

With gcc-4.4.7:
```
cc1: warnings being treated as errors
/tmp/libavif/src/colrconvert.c: In function ‘avifColorPrimariesComputeRGBToXYZD50Matrix’:
/tmp/libavif/src/colrconvert.c:109: error: passing argument 1 of ‘avifMatInv’ from incompatible pointer type
/tmp/libavif/src/colrconvert.c:25: note: expected ‘double (*)[3]’ but argument is of type ‘const double (*)[3]’
/tmp/libavif/src/colrconvert.c:118: error: passing argument 1 of ‘avifMatMul’ from incompatible pointer type
/tmp/libavif/src/colrconvert.c:48: note: expected ‘double (*)[3]’ but argument is of type ‘const double (*)[3]’
/tmp/libavif/src/colrconvert.c:124: error: passing argument 1 of ‘avifVecMul’ from incompatible pointer type
/tmp/libavif/src/colrconvert.c:76: note: expected ‘double (*)[3]’ but argument is of type ‘const double (*)[3]’
/tmp/libavif/src/colrconvert.c:136: error: passing argument 2 of ‘avifMatMul’ from incompatible pointer type
/tmp/libavif/src/colrconvert.c:48: note: expected ‘double (*)[3]’ but argument is of type ‘const double (*)[3]’
/tmp/libavif/src/colrconvert.c:139: error: passing argument 1 of ‘avifMatInv’ from incompatible pointer type
/tmp/libavif/src/colrconvert.c:25: note: expected ‘double (*)[3]’ but argument is of type ‘const double (*)[3]’
/tmp/libavif/src/colrconvert.c: In function ‘avifLinearRGBConvertColorSpace’:
/tmp/libavif/src/colrconvert.c:182: error: passing argument 1 of ‘avifVecMul’ from incompatible pointer type
/tmp/libavif/src/colrconvert.c:76: note: expected ‘double (*)[3]’ but argument is of type ‘const double (*)[3]’
```

With gcc-4.9.4:
```
In file included from /tmp/libavif/src/colrconvert.c:4:0:
/tmp/libavif/src/colrconvert.c: In function 'avifColorPrimariesComputeRGBToXYZD50Matrix':
/tmp/libavif/src/colrconvert.c:109:27: error: passing argument 1 of 'avifMatInv' from incompatible pointer type [-Werror]
     AVIF_CHECK(avifMatInv(rgbPrimaries, rgbPrimariesInv));
                           ^
/tmp/libavif/include/avif/internal.h:37:15: note: in definition of macro 'AVIF_CHECK'
         if (!(A)) {             \
               ^
/tmp/libavif/src/colrconvert.c:25:17: note: expected 'double (*)[3]' but argument is of type 'const double (*)[3]'
 static avifBool avifMatInv(double M[3][3], double I[3][3])
                 ^
/tmp/libavif/src/colrconvert.c:118:16: error: passing argument 1 of 'avifMatMul' from incompatible pointer type [-Werror]
     avifMatMul(rgbPrimaries, rgbCoefficientsMat, rgbXYZ);
                ^
/tmp/libavif/src/colrconvert.c:48:13: note: expected 'double (*)[3]' but argument is of type 'const double (*)[3]'
 static void avifMatMul(double A[3][3], double B[3][3], double C[3][3])
             ^
/tmp/libavif/src/colrconvert.c:124:16: error: passing argument 1 of 'avifVecMul' from incompatible pointer type [-Werror]
     avifVecMul(avifBradford, whitePointXYZ, lms);
                ^
/tmp/libavif/src/colrconvert.c:76:13: note: expected 'double (*)[3]' but argument is of type 'const double (*)[3]'
 static void avifVecMul(double M[3][3], const double x[3], double y[3])
             ^
/tmp/libavif/src/colrconvert.c:136:28: error: passing argument 2 of 'avifMatMul' from incompatible pointer type [-Werror]
     avifMatMul(adaptation, avifBradford, tmp);
                            ^
/tmp/libavif/src/colrconvert.c:48:13: note: expected 'double (*)[3]' but argument is of type 'const double (*)[3]'
 static void avifMatMul(double A[3][3], double B[3][3], double C[3][3])
             ^
/tmp/libavif/src/colrconvert.c:139:21: error: passing argument 1 of 'avifMatInv' from incompatible pointer type [-Werror]
     if (!avifMatInv(avifBradford, bradfordInv)) {
                     ^
/tmp/libavif/src/colrconvert.c:25:17: note: expected 'double (*)[3]' but argument is of type 'const double (*)[3]'
 static avifBool avifMatInv(double M[3][3], double I[3][3])
                 ^
/tmp/libavif/src/colrconvert.c: In function 'avifLinearRGBConvertColorSpace':
/tmp/libavif/src/colrconvert.c:182:16: error: passing argument 1 of 'avifVecMul' from incompatible pointer type [-Werror]
     avifVecMul(coeffs, rgbDouble, converted);
                ^
/tmp/libavif/src/colrconvert.c:76:13: note: expected 'double (*)[3]' but argument is of type 'const double (*)[3]'
 static void avifVecMul(double M[3][3], const double x[3], double y[3])
             ^
cc1: all warnings being treated as errors
```

**Comment by vrabaud** [comment]
@sezero , I updated the PR. Can you please try again? And comment there for any new defects. Thx

**Comment by sezero** [comment]
> @sezero , I updated the PR. Can you please try again? And comment there for any new defects. Thx

Yes: building with gcc 4.4.7 and 4.9.4 works with #1946 applied. Thanks.

**Comment by sezero** [comment]
I guess this can be closed since #1946 is in?
